### PR TITLE
3d vision msgs

### PR DESCRIPTION
**Pull request by ijnek**
This PR adds 3d vision msgs, and extracts some vision attributes out to a new msg package called soccer_vision_attribute_msgs.

Diagrams below show how these messages would fit into a 2d and 3d vision system.

soccer_3d_vision_msgs drawio

image

### CODE REVIEW

File: soccer_vision_2d_msgs/README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
-# soccer_vision_msgs
+# soccer_vision_2d_msgs
 
-This package provides messages for visually detected objects on a soccer field, such as balls, field lines and robots.
+This package provides 2d messages for visually detected objects on a soccer field, such as balls, field lines and robots.
```
Comment (jaagut) [inline soccer_vision_2d_msgs/README.md]:
```suggestion
This package provides 2d messages for visually detected objects on a soccer field, such as balls, field lines (markings) and robots.
```

File: soccer_vision_2d_msgs/README.md [hunk 2]
```diff
@@ -20,7 +20,7 @@ This package provides messages for visually detected objects on a soccer field,
 
 ## Examples
 
-![](images/soccer_vision_msgs_visualized.png)
+![](images/soccer_vision_2d_msgs_visualized.png)
```
Comment (jaagut) [inline soccer_vision_2d_msgs/README.md:23]: I believe the filename does not contain the `2d` part

File: soccer_vision_attribute_msgs/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+This package provides messages for attributes of visually detected objects on a soccer field, such as balls, field lines and robots
+
+## Messages (.msg)
+
+* Confidence
+* Goalpost
+* Robot
+
+## Examples
+
```
Comment (jaagut) [inline soccer_vision_attribute_msgs/README.md]: ...

File: soccer_vision_attribute_msgs/msg/Confidence.msg [hunk 4]
```diff
@@ -0,0 +1,6 @@
+float32 CONFIDENCE_UNKNOWN=-1
+
+# A confidence rating in the range [0-1], or -1(CONFIDENCE_UNKNOWN) by default.
+# 0 - detection module is 0% sure
+# 1 - detection module is 100% sure
+float32 confidence -1
```
Comment (jaagut) [inline soccer_vision_attribute_msgs/msg/Confidence.msg]:
```suggestion
float32 CONFIDENCE_UNKNOWN=-1.0

# A confidence rating in the range [0.0 - 1.0], or -1.0(CONFIDENCE_UNKNOWN) by default.
# 0.0 - detection module is 0% sure
# 1.0 - detection module is 100% sure
float32 confidence -1.0
```
This is not necessary, but might improve the understanding. I hope it works, haven't tested it...

File: soccer_vision_3d_msgs/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+* MarkingIntersection
+* MarkingSegment
+* MarkingArray
+* Obstacle
+* ObstacleArray
+* Robot
+* RobotArray 
+
+## Examples
+
```
Comment (jaagut) [inline soccer_vision_3d_msgs/README.md]: ...

File: soccer_vision_3d_msgs/msg/BallArray.msg [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# An array of balls with a header for global reference.
+# This msg type is used to publish all balls detected in an image.
+
+std_msgs/Header header
+soccer_vision_3d_msgs/Ball[]  balls
```
Comment (jaagut) [inline soccer_vision_3d_msgs/msg/BallArray.msg]:
```suggestion
soccer_vision_3d_msgs/Ball[] balls
```

File: soccer_vision_3d_msgs/msg/BallArray.msg [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# An array of balls with a header for global reference.
+# This msg type is used to publish all balls detected in an image.
```
Comment (jaagut) [inline soccer_vision_3d_msgs/msg/BallArray.msg]: `detected in an image` I guess it is still the case, but it sounds a little bit odd...

File: soccer_vision_3d_msgs/msg/MarkingSegment.msg [hunk 8]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (jaagut) [inline soccer_vision_3d_msgs/msg/MarkingSegment.msg]:
```suggestion
```

File: soccer_vision_3d_msgs/package.xml [hunk 9]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>soccer_vision_3d_msgs</name>
+  <version>0.0.0</version>
+  <description>A package containing some vision related message definitions in the soccer domain.</description>
```
Comment (jaagut) [inline soccer_vision_3d_msgs/package.xml]: Maybe we include a note on 3-dimensionality in there. (Probably also applies to other package.xml files)

File: soccer_vision_attribute_msgs/README.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# soccer_vision_attribute_msgs
+
+This package provides messages for attributes of visually detected objects on a soccer field, such as balls, field lines and robots
```
Comment (jaagut) [inline soccer_vision_attribute_msgs/README.md]:
```suggestion
This package provides messages for attributes of visually detected objects on a soccer field, such as balls, field lines and robots.
```
I don't know how to phrase it better, but at first I had a hard understanding, what this whole package was far. Now I know better, but it was not intuitive.

**Review by jaagut — APPROVED** [review]
Thank you for your efforts!
I have only a few notes.

Perhaps, we could add some "high level" overview of the different packages and their purpose in the root readme file

File: soccer_vision_2d_msgs/README.md (hunk 2 above)
Comment (ijnek) [reply soccer_vision_2d_msgs/README.md:23]: I'll rename the file to soccer_vision_2d_msgs_visualized.png

File: soccer_vision_2d_msgs/README.md (hunk 1 above)
Comment (ijnek) [reply soccer_vision_2d_msgs/README.md]: Good pick up, lets keep it consistent with field markings.

File: soccer_vision_3d_msgs/msg/BallArray.msg (hunk 7 above)
Comment (ijnek) [reply soccer_vision_3d_msgs/msg/BallArray.msg]: I agree, I'm going to remove ``in an image``

File: soccer_vision_attribute_msgs/README.md (hunk 10 above)
Comment (ijnek) [reply soccer_vision_attribute_msgs/README.md]: I agree, I've updated to include a better description of the package.

File: soccer_vision_attribute_msgs/README.md (hunk 3 above)
Comment (ijnek) [reply soccer_vision_attribute_msgs/README.md]: I'll remove the Example sub-heading.

File: soccer_vision_3d_msgs/README.md (hunk 5 above)
Comment (ijnek) [reply soccer_vision_3d_msgs/README.md]: I'll remove the Example sub-heading.

File: soccer_vision_3d_msgs/package.xml (hunk 9 above)
Comment (ijnek) [reply soccer_vision_3d_msgs/package.xml]: Added, and updated other package.xml files too.

**Comment by ijnek** [comment]
Thanks @jaagut, impressed with your review skills! I've fixed up everything you've pointed out.

Due to a rebase, the additional commits are a bit hard to see.

**Comment by ijnek** [comment]
I've also added an overview in the root README.md as you've suggested. Going to merge this in with the one approval, and all conversations being resolved.

**Comment by ijnek** [comment]
@Mergifyio backport galactic

**Pull request merged**